### PR TITLE
Linked list

### PR DESCRIPTION
**Pull request by chamberi**

### CODE REVIEW

File: ---------- coverage: platform linux2, python 2.7.1 [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+
+----------- coverage: platform linux, python 3.5.2-final-0 -----------
+Name                  Stmts   Miss  Cover   Missing
+---------------------------------------------------
+src/client.py            27      3    89%   15-16, 36
+src/server.py            30     30     0%   3-39
+src/test_servers.py       9      0   100%
+---------------------------------------------------
+TOTAL                    66     33    50%
```
Comment (nhuntwalker) [inline ---------- coverage: platform linux2, python 2.7.1]: Why is this file here? This doesn't apply to this repo at all. Take it out.

File: src/linked_list.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    """Defining Node Class."""
+
+    def __init__(self, val, next=None):
+        """Instantiate Node Class."""
+        self.val = val
+        self.next = next
+
+
+class LinkedList(object):
+    """Defining class of LinkedList."""
```
Comment (nhuntwalker) [inline src/linked_list.py]: When writing doc strings for classes, it's a good idea to include in the doc string the methods of the class object and what they do.

File: src/linked_list.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    def __init__(self, iterable=None):
+        """Instantiate LinkedList class."""
+        self.head = None
+        self.size_of_list = 0
+        if iterable and hasattr(iterable, "__iter__"):
+            for value in iterable:
+                self.push(value)
+        elif iterable:
+            self.push(iterable)
```
Comment (nhuntwalker) [inline src/linked_list.py]: So if it's an iterable, you push every element and if it's not an iterable you push it anyway? That's not really in line with the spec, which specifically says that it should take an optional iterable on instantiation.

File: src/linked_list.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+
+class LinkedList(object):
+    """Defining class of LinkedList."""
+
+    def __init__(self, iterable=None):
+        """Instantiate LinkedList class."""
+        self.head = None
+        self.size_of_list = 0
+        if iterable and hasattr(iterable, "__iter__"):
```
Comment (nhuntwalker) [inline src/linked_list.py]: Note that in Python 2, strings don't have the `__iter__` attribute.

File: src/linked_list.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        """Return the val of the node when searched."""
+        current = self.head
+        while current:
+            if current.val == val:
+                return current
+            current = current.next
+
+    def remove(self, node_to_delete):
+        """Remove the node passed into the parameter."""
+        current_node = self.search(node_to_delete)
```
Comment (nhuntwalker) [inline src/linked_list.py]: Remember that `remove` takes a node as a parameter, but `search` takes a value. This won't work.

File: src/linked_list.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        found = False
+        print(self.size_of_list, "Old Size of List")
```
Comment (nhuntwalker) [inline src/linked_list.py]: Print statements like these don't belong in production unless asked for.

File: src/linked_list.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        previous = self.head
+        search_node = self.head.next
+        while search_node and found is False:
```
Comment (nhuntwalker) [inline src/linked_list.py]: `...and not found`

File: src/linked_list.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            if current_node is self.head:
+                self.pop()
+                found = True
+            elif current_node.val == node_to_delete:
+                previous.next = current_node.next
+                found = True
```
Comment (nhuntwalker) [inline src/linked_list.py]: And what if you don't find the node? You just stay on the same search_node and never move forward.

File: src/linked_list.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        self.size_of_list -= 1
+        print(self.size_of_list, "New Size of List")
+        return previous.next
```
Comment (nhuntwalker) [inline src/linked_list.py]: This method isn't supposed to return anything.

File: src/linked_list.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+    def search(self, val):
+        """Return the val of the node when searched."""
+        current = self.head
+        while current:
+            if current.val == val:
+                return current
+            current = current.next
+
+    def remove(self, node_to_delete):
```
Comment (nhuntwalker) [inline src/linked_list.py:64]: If you try to remove a node that isn't an item in the list, there should be an exception raised.

File: src/test_linked_list.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
+"""This tests the linked_list module."""
+import pytest
+
+INSTANTIATION_TABLE = [
+    ("new_linked_list"),
+]
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: If your parametrization table only has one item in it, it's not worth having the table.

File: src/test_linked_list.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+REMOVE_TABLE = [
+    (3, 7, 3, 2, 7),
+    (7, 8, 7, 5, 8),
+    (5, 8, 7, 5, 7),
+]
+
+
+@pytest.mark.parametrize("list_name", INSTANTIATION_TABLE)
+def test_linkedlist(list_name):
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: There's no need to have a parameter in this function declaration since you just use that name to instantiate your linked list

File: src/test_linked_list.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    list_name = LinkedList()
+    assert list_name.size_of_list == 0
+
+
+@pytest.mark.parametrize("val, result", PUSH_TABLE)
+def test_push(val, result):
+    """Test push method to add value to front of linked list."""
+    from linked_list import LinkedList
+    list_name = LinkedList(val)
+    assert list_name.head.val == result
```
Comment (nhuntwalker) [inline src/test_linked_list.py:67]: Both tests coming from your `PUSH_TABLE` are the same test twice: "what happens if I only push on one value?". Try pushing on more than one.

File: src/test_linked_list.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+
+@pytest.mark.parametrize("val1, val2", POP_TABLE)
+def test_pop(val1, val2):
+    """Test pop method to remove node off the front of a linked list."""
+    from linked_list import LinkedList
+    list_name = LinkedList(val1)
+    list_name.push(val2)
+    assert list_name.pop() == val2
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: Same comment here as above. Also, and this speaks to a different comment I made in your main code file. Your list shouldn't take a regular old value at instantiation, it should take an iterable.

File: src/test_linked_list.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+
+@pytest.mark.parametrize("val1, result", SIZE_TABLE)
+def test_size(val1, result):
+    """Test size method in LinkedList class."""
+    from linked_list import LinkedList
+    list_name = LinkedList(val1)
+    assert list_name.size() == result
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: Just for conciseness, you shouldn't need a parameter for the actual length of the thing you're passing in. You could just test that the size of your list matches the result of `len(val1)`

File: src/test_linked_list.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+
+@pytest.mark.parametrize('val_to_search_for, val1, val2, val3, result', SEARCH_TABLE)
+def test_search(val_to_search_for, val1, val2, val3, result):
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: Why do you need three separate values to push? Why not just pass in an iterable containing those values?

File: src/test_linked_list.py [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+@pytest.mark.parametrize("val_to_delete, val1, val2, val3, result", REMOVE_TABLE)
+def test_remove(val_to_delete, val1, val2, val3, result):
+    """Test remove method in LinkedList class."""
+    from linked_list import LinkedList
+    list_name = LinkedList(val1)
+    list_name.push(val2)
+    list_name.push(val3)
+    assert_result = list_name.remove(val_to_delete)
+    assert assert_result.val == result
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: This isn't the way to test this method. Test that the thing no longer exists in your list, or test that the size of the list decreases or both. Also get some variety in your test for edge cases. What if you remove the head of the list? What about the tail?

File: src/test_linked_list.py [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def test_remove_last():
+    """Test for removing the last value."""
+    from linked_list import LinkedList
+    list_name = LinkedList(5)
+    list_name.push(6)
+    list_name.push(7)
+    assert list_name.remove(5) is None
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: Again, not the way to test this method.

File: src/test_linked_list.py [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def test_display():
+    """Test for displaying as a tuple."""
+    from linked_list import LinkedList
+    list_name = LinkedList(5)
+    list_name.push(6)
+    list_name.push('*adf')
+    assert list_name.display() == '(*adf, 6, 5, )'
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: Your last item in the `display` method shouldn't be followed by a hanging comma.

File: ---------- coverage: platform linux2, python 2.7.1 (hunk 1 above)
Comment (Copenbacon) [reply ---------- coverage: platform linux2, python 2.7.1]: removed

File: src/linked_list.py (hunk 4 above)
Comment (Copenbacon) [reply src/linked_list.py]: updated

File: src/linked_list.py (hunk 3 above)
Comment (Copenbacon) [reply src/linked_list.py]: updated

File: src/linked_list.py (hunk 10 above)
Comment (Copenbacon) [reply src/linked_list.py:64]: updated

File: src/linked_list.py (hunk 5 above)
Comment (Copenbacon) [reply src/linked_list.py]: updated

File: src/linked_list.py (hunk 6 above)
Comment (Copenbacon) [reply src/linked_list.py]: removed

File: src/linked_list.py (hunk 7 above)
Comment (Copenbacon) [reply src/linked_list.py]: updated

File: src/linked_list.py (hunk 8 above)
Comment (Copenbacon) [reply src/linked_list.py]: updated

File: src/linked_list.py (hunk 9 above)
Comment (Copenbacon) [reply src/linked_list.py]: removed

File: src/test_linked_list.py (hunk 11 above)
Comment (Copenbacon) [reply src/test_linked_list.py]: removed

File: src/test_linked_list.py (hunk 12 above)
Comment (Copenbacon) [reply src/test_linked_list.py]: removed

File: src/test_linked_list.py (hunk 13 above)
Comment (Copenbacon) [reply src/test_linked_list.py:67]: updated/changed test completely. See INSTANTIATION_TABLE in test file

File: src/test_linked_list.py (hunk 16 above)
Comment (Copenbacon) [reply src/test_linked_list.py]: fixed

File: README.md [hunk 20]
```diff
@@ -1 +1,80 @@
...
+py35 installed: coverage==4.2,data-structures==0.1,decorator==4.0.10,ipython==5.1.0,ipython-genutils==0.1.0,pexpect==4.2.1,pickleshare==0.7.4,prompt-toolkit==1.0.9,ptyprocess==0.5.1,py==1.4.31,Pygments==2.1.3,pytest==3.0.5,pytest-cov==2.4.0,simplegeneric==0.8.1,six==1.10.0,traitlets==4.3.1,wcwidth==0.1.7
+py35 runtests: PYTHONHASHSEED='1421974174'
+py35 runtests: commands[0] | py.test src --cov=src --cov-report term-missing
+=================================== test session starts ====================================
+platform linux -- Python 3.5.2, pytest-3.0.5, py-1.4.31, pluggy-0.4.0
+rootdir: /media/copenbacon/OS/Users/Conor/codefellows/401/data-structures, inifile: 
+plugins: cov-2.4.0
+collected 24 items 
+
+src/test_linked_list.py ........................
```
Comment (nhuntwalker) [inline README.md]: I don't need the full output text from pytest or tox. I just want the coverage reports.

File: setup.py [hunk 21]
```diff
@@ -0,0 +1,18 @@
+"""Setup.py for data-structures repo."""
+
+from setuptools import setup
+
+setup(
+    name="data-structures",
+    description="Implementation of a Linked List.",
```
Comment (nhuntwalker) [inline setup.py]: This description doesn't apply to the total repository

File: src/linked_list.py [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+class LinkedList(object):
+    """Defining class of LinkedList. 
+
+    push(val): adds a node at the head, increments size of list.
+    pop(): removes the node at the head, deincrements size of list.
+    size(): returns size of LinkedList
+    search(val): searches list for node containing val and returns that node.
+    remove(node_to_delete): takes a node as argument, searches for it, and removes it from list, then deincrements size of list
+    display(): returns LinkedList as a stringified tuple-looking object.
+
```
Comment (nhuntwalker) [inline src/linked_list.py]: Don't need this white space.

File: src/linked_list.py [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+    def __init__(self, iterable=None):
+        """Instantiate LinkedList class."""
+        self.head = None
+        self.size_of_list = 0
+        if iterable and hasattr(iterable, "__iter__") or isinstance(iterable, str):
+            for value in iterable:
+                self.push(value)
+        elif iterable:
+            raise TypeError
+            print("Please submit an iterable when instantiating the list. You can use the push method to add one value at a time.")
```
Comment (nhuntwalker) [inline src/linked_list.py]: You can include this message as a part of the raised exception. Like so:
```python
raise TypeError("Please submit an iterable when instantiating the list. You can use the push method to add one value at a time.")
```

File: src/linked_list.py [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+        """Return the val of the node when searched."""
+        current = self.head
+        while current:
+            if current.val == val:
+                return current
+            current = current.next
+        if current is None:
+            raise NameError
+            print("Parameter given, {}, is not in list".format(str(val)))
+        return current
```
Comment (nhuntwalker) [inline src/linked_list.py]: Why do you need lines 64 through 67? If search doesn't find what it's looking for, it should just return None. It's in the spec.

"search(val) will return the node containing ‘val’ in the list, if present, else None"

File: src/linked_list.py [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+        while search_node and not found:
+            if current_node is self.head:
+                self.pop()
+                found = True
+            elif current_node.val == node_to_delete.val:
+                previous.next = current_node.next
+                found = True
+                self.size_of_list -= 1
+            else:
+                raise NameError
```
Comment (nhuntwalker) [inline src/linked_list.py]: Why are you raising a NameError? Name Errors are for when you try to use a variable that wasn't yet declared. This would be more like a ValueError

File: src/linked_list.py [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+            print("Parameter given, {}, is not in list".format(str(val)))
+        return current
+
+    def remove(self, node_to_delete):
+        """Remove the node passed into the parameter."""
+        current_node = self.search(node_to_delete.val)
+        found = False
+        previous = self.head
+        search_node = self.head.next
+        while search_node and not found:
```
Comment (nhuntwalker) [inline src/linked_list.py]: You don't actually walk through the list anywhere. You need to iterate through the list by pointing `search_node` at the next available node.

File: src/linked_list.py [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+            if current_node is self.head:
+                self.pop()
+                found = True
+            elif current_node.val == node_to_delete.val:
```
Comment (nhuntwalker) [inline src/linked_list.py]: This has nothing to do with `search_node`

File: README.md [hunk 28]
```diff
@@ -1 +1,83 @@
-# data-structures
\ No newline at end of file
+# data-structures
```
Comment (nhuntwalker) [inline README.md:1]: It's great that you're showing your code coverage in your README. Truly, it is. But your README is otherwise useless for someone who wants to actually *use* your code moving forward. Describe the modules encapsulated by your package

File: src/linked_list.py [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+
+    def __init__(self, iterable=None):
+        """Instantiate LinkedList class."""
+        self.head = None
+        self.size_of_list = 0
+        if iterable and hasattr(iterable, "__iter__") or isinstance(iterable, str):
+            for value in iterable:
+                self.push(value)
+        elif iterable:
+            raise TypeError("Please submit an iterable when instantiating the list. You can use the push method to add one value at a time.")
```
Comment (nhuntwalker) [inline src/linked_list.py:35]: Love this.

File: src/linked_list.py [hunk 30]
```diff
@@ -0,0 +1,95 @@
...
+        return self.size_of_list
+
+    def search(self, val):
+        """Return the val of the node when searched."""
+        current = self.head
+        while current:
+            if current.val == val:
+                return current
+            current = current.next
+        return current
```
Comment (nhuntwalker) [inline src/linked_list.py]: You actually don't need this here. Functions without return statements will return `None` by default. Let Python do the work for you.

File: src/linked_list.py [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+        except AttributeError:
+            current_node = node_to_delete
+        if current_node is None:
+            raise ValueError("Node not in list")
+        search_node = self.head
+        found = False
+        while not found:
+            if current_node is self.head:
+                self.pop()
+                found = True
```
Comment (nhuntwalker) [inline src/linked_list.py]: or you could just `break` here.

File: src/test_linked_list.py [hunk 32]
```diff
@@ -0,0 +1,136 @@
...
+    assert search_list.search(9) is None
+
+
+@pytest.mark.parametrize("val_to_delete, val1, result", REMOVE_TABLE)
+def test_remove(val_to_delete, val1, result):
+    """Test remove method in LinkedList class."""
+    from linked_list import LinkedList
+    list_name = LinkedList(val1)
+    list_name.remove(list_name.search(val_to_delete))
+    assert list_name.size() == len(val1) - 1
```
Comment (nhuntwalker) [inline src/test_linked_list.py:107]: Check that not only does the size decrease, but the previous node points to the appropriate next node.

File: src/test_linked_list.py [hunk 33]
```diff
@@ -0,0 +1,136 @@
...
+
+
+def test_remove_last():
+    """Test for removing the last value."""
+    from linked_list import LinkedList
+    list_name = LinkedList([5, 6, 7])
+    list_name.remove(list_name.search(5))
+    assert list_name.size() == 2
```
Comment (nhuntwalker) [inline src/test_linked_list.py:124]: More than just the size change, this should test that the node coming before the last node now points to None

File: src/test_linked_list.py [hunk 34]
```diff
@@ -0,0 +1,136 @@
...
+
+
+def test_remove_head():
+    """Test for removing the last value."""
+    from linked_list import LinkedList
+    list_name = LinkedList([5, 6, 7])
+    list_name.remove(list_name.search(7))
+    assert list_name.size() == 2
```
Comment (nhuntwalker) [inline src/test_linked_list.py:141]: This should test that a new `head` was created.
